s3repo/model.py

```diff
@@ -208,6 +181,177 @@ def _upload_files(self, package, tarantool_series, origin_files):
+        # contains files, we request if objects with this prefix("path") exist
+        # (it is enough for us to know that the first level content is exists).
+        #
+        # See https://github.com/boto/boto3/issues/134 for how to list first
+        # level content by a specific prefix.
+
+        path = base_path + '/'
+        list_objs = self.bucket.meta.client.list_objects_v2(
+            Bucket=self.bucket.name,
+            Delimiter='/',
+            Prefix=path
+        )
+        if list_objs.get('CommonPrefixes') is None:
+            return []
+
+        # In the case of a deb-base distribution, the meta information about
+        # packages in all versions of the distribution is updated together.
+        # So we just return the path to the distribution
+        return [path]
+
+    def _get_rpm_repo_path(self, base_path, dist_versions):
+        """Returns the list of the paths to the rpm-based reposies
+        for updating metainformation with the 'mkrepo' tool.
+        base_path(string) - path to the distribution.
+        dist_versions(list) - list of the distribution versions.
+        """
+        # See the first comment in "_get_deb_repo_path" method.
+
+        # In the case of an rpm-based distribution, one distribution version can
+        # include several repositories (for example "x86_64" and "SRPM").
+        # We must collect them all.
+        repos_list = []
+        for ver in dist_versions:
+            # Path to all repositories of the distribution version.
+            common_path = '/'.join([base_path, ver]) + '/'
+            list_objs = self.bucket.meta.client.list_objects_v2(
+                Bucket=self.bucket.name,
+                Delimiter='/',
+                Prefix=common_path
+            )
+            dist_repos_list = list_objs.get('CommonPrefixes')
+            if dist_repos_list is None:
+                continue
+            for repo in dist_repos_list:
+                # In this context, "Prefix" is a path to the repository.
+                repos_list.append(repo['Prefix'])
+
+        return repos_list
+
+    def _get_repository_list(self):
+        """Returns a list of paths to repositories in the current bucket."""
+        supported_repos = self.get_supported_repos()
+        repos_list = []
+        result_list = []
+
+        # Since collecting the list of paths to repositories involves a large
+        # number of S3 requests through the network, it is recommended to use a
+        # thread pool to execute them in parallel (the thread will be idle for a
+        # "long" time, waiting for a response from S3).
+        #
+        # The number of processes = 20 was chosen experimentally.
+        with ThreadPool(processes=20) as pool:
+            for kind in supported_repos['repo_kind']:
+                for series in supported_repos['tarantool_series']:
+                    for dist, dist_description in supported_repos['distrs'].items():
+                        path = ''
```

Comment:
Why do you add default value if all your conditional branches have path definitions? I think mypy have not to raise a warning in this situation.

Reply:
Yeah.
There are two reasons for this:
- (The main) Me and other programmers of our team came from the "C world" and in "C" the scope of the variable is limited to the condition branch. I agree that in python it is possible and I do not try to drag the techniques of writing code from one language to another, but the option of declaring a variable inside a branch, and then using it outside greatly confuses me (and I think not only) and complicates the reading of the code.
- We can get an error later if a new one branch is added where `path` was not declared.